samza-core/src/main/java/org/apache/samza/processorgraph/ProcessorGraph.java

```diff
@@ -0,0 +1,299 @@
+          throw new IllegalArgumentException(
+              String.format("Internal stream %s should have both producers and consumers", edge.getFormattedSystemStream()));
+        }
+      });
+  }
+
+  /**
+   * Validate all nodes are reachable by sources.
+   */
+  private void validateReachability() {
+    // validate all nodes are reachable from the sources
+    final Set<ProcessorNode> reachable = findReachable();
+    if (reachable.size() != nodes.size()) {
+      Set<ProcessorNode> unreachable = new HashSet<>(nodes.values());
+      unreachable.removeAll(reachable);
+      throw new IllegalArgumentException(String.format("Processors %s cannot be reached from Sources.",
+          String.join(", ", unreachable.stream().map(ProcessorNode::getId).collect(Collectors.toList()))));
+    }
+  }
+
+  /**
+   * Find the reachable set of nodes using BFS.
+   * Package private for test.
+   * @return reachable set of nodes
+   */
+  Set<ProcessorNode> findReachable() {
+    Queue<ProcessorNode> queue = new ArrayDeque<>();
+    Set<ProcessorNode> visited = new HashSet<>();
+
+    sources.forEach(source -> {
+        List<ProcessorNode> next = source.getTargetNodes();
+        queue.addAll(next);
+        visited.addAll(next);
+      });
+
+    while (!queue.isEmpty()) {
+      ProcessorNode node = queue.poll();
+      node.getOutEdges().stream().flatMap(edge -> edge.getTargetNodes().stream()).forEach(target -> {
+          if (!visited.contains(target)) {
+            visited.add(target);
+            queue.offer(target);
+          }
+        });
+    }
+
+    return visited;
+  }
+
+  /**
+   * An variation of Kahn's algorithm of topological sorting.
+   * This algorithm also takes account of the simple loops in the graph
+   * Package private for test.
+   * @return topologically sorted ProcessorNode(s)
+   */
+  List<ProcessorNode> topologicalSort() {
+    Collection<ProcessorNode> pnodes = nodes.values();
+    Queue<ProcessorNode> q = new ArrayDeque<>();
+    Map<String, Long> indegree = new HashMap<>();
+    Set<ProcessorNode> visited = new HashSet<>();
+    pnodes.forEach(node -> {
+        String nid = node.getId();
+        //only count the degrees of intermediate streams since sources have degree 0
+        long degree = node.getInEdges().stream().filter(e -> !sources.contains(e)).count();
+        indegree.put(nid, degree);
+
+        if (degree == 0L) {
+          // start from the nodes that only consume from sources
+          q.add(node);
+          visited.add(node);
+        }
+      });
+
+    List<ProcessorNode> sortedNodes = new ArrayList<>();
+    Set<ProcessorNode> reachable = new HashSet<>();
+    while (sortedNodes.size() < pnodes.size()) {
+      while (!q.isEmpty()) {
+        ProcessorNode node = q.poll();
+        sortedNodes.add(node);
+        node.getOutEdges().stream().flatMap(edge -> edge.getTargetNodes().stream()).forEach(n -> {
+            String nid = n.getId();
+            Long degree = indegree.get(nid) - 1;
```

Comment:
Yup, the logic looks good. I meant to say that it'll be clearer to keep a running count and subtract that from the indegree, rather than mutating the indegree map directly.